### PR TITLE
Move plotting functions to separate modules

### PR DESCRIPTION
**Pull request by willschlitzer**
As mentioned in #685, this is to split up `base_plotting.py` and move the plotting functions to separate scripts, following the method used by @weiji14 in #686.

**Reminders**

- [x] Run `make format` and `make check` to make sure the code follows the style guide.
- [x] Add tests for new features or tests that would have caught the bug that you're fixing.

**Notes**

- You can write `/format` in the first line of a comment to lint the code automatically

### CODE REVIEW

**Comment by willschlitzer** [comment]
@GenericMappingTools/python I moved all of the plotting functions to their own modules. I'm running into a few style errors that I'm not sure how to address.

`pygmt/src/contour.py:98:13: W0212: Access to a protected member _preprocess of a client class (protected-access)`
I'm guessing this is because the functions use `_preprocess` but don't import it, but the tests successfully running makes me assume that it works because `_preprocess` is in the class that these modules are imported into.

`pygmt/base_plotting.py:8:0: R0903: Too few public methods (0/2) (too-few-public-methods)`
I'm assuming this is because the class doesn't really "do" all that much specifically, since the functions are all housed in their own modules. I'm assuming this is acceptable, otherwise functions would just be arbitrarily moved back to `base_plotting.py`.

`pygmt/src/text.py:48:4: W0621: Redefining name 'text' from outer scope (line 40) (redefined-outer-name)`
I'm guessing "text" is a common/special word, so having it in the name of both a function and module is causing difficulties. Not sure how it should be addressed.

File: pygmt/src/text.py
Comment (seisman) [inline pygmt/src/text.py:2]:
```suggestion
text - Plot text on a figure.
```

File: pygmt/src/basemap.py
Comment (seisman) [inline pygmt/src/basemap.py:3]: See line 36 of this file. It would be better if the function description is the same as the one shown in the API reference (i.e., line 36).

Comment (willschlitzer) [reply pygmt/src/basemap.py]: Copied line 36.

**Comment by seisman** [comment]
> @GenericMappingTools/python I moved all of the plotting functions to their own modules. 

Great work!

> `pygmt/src/contour.py:98:13: W0212: Access to a protected member _preprocess of a client class (protected-access)`
> I'm guessing this is because the functions use `_preprocess` but don't import it, but the tests successfully running makes me assume that it works because `_preprocess` is in the class that these modules are imported into.

Functions with a leading underscore mean they're protected, and _should_ not be accessed outside the class. This function does nothing, and it should be safe to remove it from all functions. I'm not sure if we still need it in the future.
 
> `pygmt/base_plotting.py:8:0: R0903: Too few public methods (0/2) (too-few-public-methods)`
> I'm assuming this is because the class doesn't really "do" all that much specifically, since the functions are all housed in their own modules. I'm assuming this is acceptable, otherwise functions would just be arbitrarily moved back to `base_plotting.py`.

OK to disable the warning.
 
> `pygmt/src/text.py:48:4: W0621: Redefining name 'text' from outer scope (line 40) (redefined-outer-name)`
> I'm guessing "text" is a common/special word, so having it in the name of both a function and module is causing difficulties. Not sure how it should be addressed.

The function is called `text`. Can we also use `text` in the arguments? 😕 
```
def text(
    self,
    textfiles=None,
    x=None,
    y=None,
    position=None,
    text=None,
    angle=None,
    font=None,
    justify=None,
    **kwargs,
):
```

**Comment by willschlitzer** [comment]
@seisman So should I change the function name or the argument name?

**Comment by willschlitzer** [comment]
@seisman When I try removing `_preprocess` all of my tests fail. I get:

`pygmt.exceptions.GMTCLibError: Module 'psconvert' failed with status code 78`

I can't tell what it does, but it has some effect.

**Comment by seisman** [comment]
> @seisman So should I change the function name or the argument name?

Need more thoughts and comments from @GenericMappingTools/python-maintainers


> @seisman When I try removing `_preprocess` all of my tests fail. I get:
> 
> `pygmt.exceptions.GMTCLibError: Module 'psconvert' failed with status code 78`
> 
> I can't tell what it does, but it has some effect.

I think you can't remove `_preprocess` from the functions in `pygmt/figure.py`.

**Comment by weiji14** [comment]
> > @seisman So should I change the function name or the argument name?
> 
> Need more thoughts and comments from @GenericMappingTools/python-maintainers

Maybe just keep the `text` function in `base_plotting.py` for now :slightly_smiling_face:

> I think you can't remove `_preprocess` from the functions in `pygmt/figure.py`.

I feel like the `_preprocess` function might be needed for some obscure reason, e.g. when people do something like:

```python
fig1 = pygmt.Figure()
fig2 = pygmt.Figure()

fig2.basemap()
fig1.basemap()

fig1.show()
fig2.show()
```

Do things get passed to the correct figure in this case?

**Comment by willschlitzer** [comment]
@GenericMappingTools/python Could I get additional feedback/comments for this pull request?

**Comment by seisman** [comment]
> I feel like the `_preprocess` function might be needed for some obscure reason, e.g. when people do something like:
> 
> ```python
> fig1 = pygmt.Figure()
> fig2 = pygmt.Figure()
> 
> fig2.basemap()
> fig1.basemap()
> 
> fig1.show()
> fig2.show()
> ```
> 
> Do things get passed to the correct figure in this case?

Can you remove `_preprocess` from `basemap()` and try if the above example works? I think it doesn't matter.

File: pygmt/base_plotting.py
Comment (weiji14) [inline pygmt/base_plotting.py:56]: You know, if we're going to move **every** plotting function to the `src/` folder, we might as well delete `base_plotting.py` and import everything in `figure.py` instead. Or we could keep a **few** core functions in `base_plotting.py` (e.g. `basemap`, and `inset` perhaps as mentioned in https://github.com/GenericMappingTools/pygmt/pull/788#issuecomment-766760050). What do people think?

```suggestion
from pygmt.src import (  # pylint: disable=import-outside-toplevel
        basemap,
        coast,
        colorbar,
        contour,
        grdcontour,
        grdimage,
        grdview,
        image,
        legend,
        logo,
        meca,
        plot,
        plot3d,
        text,
    )
```

Comment (seisman) [reply pygmt/base_plotting.py:56]: > You know, if we're going to move **every** plotting function to the `src/` folder, we might as well delete `base_plotting.py` and import everything in `figure.py` instead. 

I like this one.

Comment (willschlitzer) [reply pygmt/base_plotting.py:56]: I committed this changed in 76c4969 but it led to 191 tests failing when I ran `make test`. I'm not sure what the issue is, but I've reverted for now.

**Comment by willschlitzer** [comment]
> > I feel like the `_preprocess` function might be needed for some obscure reason, e.g. when people do something like:
> > ```python
> > fig1 = pygmt.Figure()
> > fig2 = pygmt.Figure()
> > 
> > fig2.basemap()
> > fig1.basemap()
> > 
> > fig1.show()
> > fig2.show()
> > ```
> > 
> > 
> > Do things get passed to the correct figure in this case?
> 
> Can you remove `_preprocess` from `basemap()` and try if the above example works? I think it doesn't matter.
 ```python
 fig1 = pygmt.Figure()
fig2 = pygmt.Figure()
fig2.basemap(frame="a", region=[0, 20, 0, 20])
fig1.basemap(frame="a", region=[0, 10, 0, 10])
fig1.show()
fig2.show()
```
When I ran the above code after removing `_preprocess` I received `GMTCLibError: Module 'psconvert' failed with status code 78:
psconvert [ERROR]: No hidden PS file /home/will/.gmt/sessions/gmt_session.38138/gmt_1.ps- found` and it resulted in 10 failed tests.

File: pygmt/base_plotting.py
Comment (weiji14) [reply pygmt/base_plotting.py:56]: Oops sorry, just realized it was missing an indentation... Try this instead:

```suggestion
    from pygmt.src import (  # pylint: disable=import-outside-toplevel
        basemap,
        coast,
        colorbar,
        contour,
        grdcontour,
        grdimage,
        grdview,
        image,
        legend,
        logo,
        meca,
        plot,
        plot3d,
        text,
    )
```

Comment (weiji14) [reply pygmt/base_plotting.py]: Ok, I applied https://github.com/GenericMappingTools/pygmt/pull/808#discussion_r570775763 in 127b42874466c66eb970fb53a8846db9ff1f73e8.

File: pygmt/src/text.py
Comment (weiji14) [inline pygmt/src/text.py:42]:
```suggestion
def text_(
```

How about this to fix the pylint `W0621: Redefining name 'text' from outer scope (line 40) (redefined-outer-name)` warning? And then use `from pygmt.src.text import text_ as text` in `pygmt/src/__init__.py`.

File: pygmt/src/__init__.py
Comment (weiji14) [inline pygmt/src/__init__.py:26]:
```suggestion
from pygmt.src.text import text_ as text
```

File: pygmt/base_plotting.py
Comment (weiji14) [inline pygmt/base_plotting.py:59]: As mentioned in https://github.com/GenericMappingTools/pygmt/pull/808#discussion_r570575569, could you delete `base_plotting.py` (can't believe I'm saying this!) and move this import chunk to `figure.py`? This will fix the pylint `R0903: Too few public methods (0/2) (too-few-public-methods)` recommendation. Feel free to copy any of the docstrings if needed.

Comment (willschlitzer) [reply pygmt/base_plotting.py:59]: @weiji14 I just moved this import string to `figure.py` and removed imports for `base_plotting.py`, but when I ran `make test` I had 196 test failures. I'm not sure what the need to `_preprocess` is, but it seems to play some sort of role.

Comment (weiji14) [reply pygmt/base_plotting.py:59]: Just double checking that you placed the imports under `class Figure():`? I.e. indented 4 spaces? The `_preprocess` class function activates the correct figure to plot in.

Comment (willschlitzer) [reply pygmt/base_plotting.py]: ...nope 🤦‍♂️

Removed `base_plotting.py` in 5ac2e07 and actually put the imports in `Figure()` and it worked (or at least passed all of the tests)!

Comment (seisman) [reply pygmt/base_plotting.py]: > As mentioned in [#808 (comment)](https://github.com/GenericMappingTools/pygmt/pull/808#discussion_r570575569), could you delete `base_plotting.py` (can't believe I'm saying this!) and move this import chunk to `figure.py`? This will fix the pylint `R0903: Too few public methods (0/2) (too-few-public-methods)` recommendation. Feel free to copy any of the docstrings if needed.

We should be careful with the changes. There are two different `_preprocess()` functions in `Figure()` and `BasePlotting`. If I understand it correctly, the `_preprocess` function if `BasePlotting` override the one in `Figure()`.

Comment (willschlitzer) [reply pygmt/base_plotting.py]: @seisman The tests are passing; is there an issue with deleting base_plotting.py?

Comment (weiji14) [reply pygmt/base_plotting.py]: > We should be careful with the changes. There are two different `_preprocess()` functions in `Figure()` and `BasePlotting`. If I understand it correctly, the `_preprocess` function if `BasePlotting` override the one in `Figure()`.

I feel like the `_preprocess` in base_plotting.py was just a dummy placeholder, and the one in figure.py is the actual one used. So probably the other way around? Might need to read up on subclassing, I'm not terribly good at object orientated conventions.

**Review by weiji14 — APPROVED** [review]
I think this is good to merge now (after updating the branch), what do you think @seisman?

Deleting `base_plotting.py` and moving every module to `src` will cause conflicts in several open PRs (e.g. `velo` at #525, `triangulate` at #731, the `grdimage` work at #750, etc), but we can sort those out later :upside_down_face:

File: pygmt/base_plotting.py
Comment (seisman) [reply pygmt/base_plotting.py]: I thought `BasePlotting()` is a subclass of `Figure()`, now it makes sense to me.

File: pygmt/src/__init__.py
Comment (seisman) [inline pygmt/src/__init__.py:26]: Please add a comment explaining why `text` is different.

File: pygmt/src/text.py
Comment (seisman) [inline pygmt/src/text.py:2]:
```suggestion
text - Plot text on a figure.
```

**Review by seisman — COMMENTED** [review]
Looks good to me, except a few tiny suggestions.

**Pull request merged**
